### PR TITLE
Manage url and http exceptions for non-existent entities

### LINKED ISSUE #8: urllib.error.HTTPError: HTTP Error 400: Bad Request
>>> from wikidata.client import Client
>>> client = Client()
>>> entity = client.get('20145', load=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\ProgramData\Anaconda3\lib\site-packages\wikidata\client.py", line 139, in get
    entity.load()
  File "C:\ProgramData\Anaconda3\lib\site-packages\wikidata\entity.py", line 239, in load
    result = self.client.request(url)
  File "C:\ProgramData\Anaconda3\lib\site-packages\wikidata\client.py", line 193, in request
    response = self.opener.open(url)
  File "C:\ProgramData\Anaconda3\lib\urllib\request.py", line 532, in open
    response = meth(req, response)
  File "C:\ProgramData\Anaconda3\lib\urllib\request.py", line 642, in http_response
    'http', request, response, code, msg, hdrs)
  File "C:\ProgramData\Anaconda3\lib\urllib\request.py", line 570, in error
    return self._call_chain(*args)
  File "C:\ProgramData\Anaconda3\lib\urllib\request.py", line 504, in _call_chain
    result = func(*args)
  File "C:\ProgramData\Anaconda3\lib\urllib\request.py", line 650, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 400: Bad Request

### PR DESCRIPTION
**Pull request by k----n**
Entity does not exist so the JSON is not returned by wikidata.

When an http or url exception is raised by a bad request, the entity data is set to none.

An empty entity is returned with an id that does not exist.

Fixes #8 .

### CODE REVIEW

**Comment by k----n** [comment]
Something is wrong with $GITHUB_ACCESS_TOKEN?

CI fails for python 3.6 .

File: wikidata/client.py [hunk 1]
```diff
@@ -177,20 +178,28 @@ def decode_datavalue(self,
...
         logger = logging.getLogger(__name__ + '.Client.request')
         url = urllib.parse.urljoin(self.base_url, path)
         result = self.cache_policy.get(CacheKey(url))
         if result is None:
             logger.debug('%r: no cache; make a request...', url)
-            response = self.opener.open(url)
+            try:
+                response = self.opener.open(url)
+            except urllib.error.HTTPError as e:
+                logger.debug('HTTP error code: ', e.code)
```
Comment (dahlia) [inline wikidata/client.py]: For the sake of debug, it could log a traceback as well.  The `exc_info=True` option purposes that.  [(See also the docs.)][1]

[1]: https://docs.python.org/3/library/logging.html#logging.Logger.debug

File: wikidata/entity.py [hunk 2]
```diff
@@ -237,25 +237,28 @@ def attributes(self) -> Mapping[str, object]:
...
-            data = entities[entity_id]
-            redirected = True
-        assert isinstance(data, collections.abc.Mapping)
-        self.data = data
-        self.id = entity_id
-        if redirected:
-            canon = self.client.get(entity_id, load=False)
-            if canon.data is None:
-                canon.data = dict(data)
+        if result is not None:
```
Comment (dahlia) [inline wikidata/entity.py]: It could be more readable if it [returns early][1].  (Or at least, it's more maintainable to me.)

[1]: http://www.itamarweiss.com/personal/2018/02/28/return-early-pattern.html

File: wikidata/client.py (hunk 1 above)
Comment (k----n) [reply wikidata/client.py]: I added the argument.

File: wikidata/entity.py (hunk 2 above)
Comment (k----n) [reply wikidata/entity.py]: I made it return early.

**Comment by k----n** [comment]
- I added a test to tests/client_test.py
- Rebased with master
- Python 3.7 build fails in CI

File: wikidata/client.py [hunk 3]
```diff
@@ -177,21 +178,30 @@ def decode_datavalue(self,
...
-    def request(self, path: str) -> Union[
+    def request(self, path: str) -> Optional[Union[
         bool, int, float, str,
         Mapping[
             str,
             Union[bool, int, float, str, Mapping[str, object], Sequence]
         ],
         Sequence[Union[bool, int, float, str, Mapping[str, object], Sequence]]
-    ]:
+    ]]:
```
Comment (dahlia) [inline wikidata/client.py:188]: [According to the `typing` API docs][1]:

> `Optional[X]` is equivalent to `Union[X, None]`.

Therefore, you don't have to put `Union[A, B, ...]` inside `Optional[...]`.  Instead, we could write the equivalent code like `Union[A, B, ..., None]`.

[1]: https://docs.python.org/3/library/typing.html#typing.Optional

File: wikidata/client.py [hunk 4]
```diff
@@ -177,21 +178,30 @@ def decode_datavalue(self,
...
-            response = self.opener.open(url)
-            buffer_ = io.TextIOWrapper(response, encoding='utf-8')
+            try:
+                response = self.opener.open(url)
+            except urllib.error.HTTPError as e:
+                logger.debug('HTTP error code: ', e.code, exc_info=True)
+                return None
+            except urllib.error.URLError as e:
+                logger.debug('Failed to reach server: ', e.reason, exc_info=True)
+                return None
```
Comment (dahlia) [inline wikidata/client.py:201]: Intention behind this your change seems to make this method to return `None` if and only if a requested entity does not exist, right?  We should be able to distinguish server errors (Wikidata's fault) from non-existent entities.  However, catching every `HTTPError` and `URLError` could make Wikidata's temporary server errors to look like non-existent entities.

Therefore, this code should be adjusted to the following direction:

- Since `URLError` is too vague (it covers any kind of exceptions `urllib` can throw), I believe we should not catch this type of exceptions.
- There are two types of error codes in HTTP: 400 series (client errors) & 500 series (server errors).  And `HTTPError` represents both kinds of errors as I recall.  Therefore, we should not catch server errors here.  IMHO you'd better catch only exceptions with `e.code == 404` and re`raise` any others.

File: wikidata/client.py [hunk 5]
```diff
@@ -177,21 +178,30 @@ def decode_datavalue(self,
...
         url = urllib.parse.urljoin(self.base_url, path)
         result = self.cache_policy.get(CacheKey(url))
         if result is None:
             logger.debug('%r: no cache; make a request...', url)
-            response = self.opener.open(url)
-            buffer_ = io.TextIOWrapper(response, encoding='utf-8')
+            try:
+                response = self.opener.open(url)
+            except urllib.error.HTTPError as e:
+                logger.debug('HTTP error code: ', e.code, exc_info=True)
```
Comment (dahlia) [inline wikidata/client.py:197]: This will ignore the second argument `e.code` and the log message ends with a single space.  This should be changed like:

```suggestion
                logger.debug('HTTP error code: %s', e.code, exc_info=True)
```

File: wikidata/entity.py [hunk 6]
```diff
@@ -237,6 +237,10 @@ def attributes(self) -> Mapping[str, object]:
     def load(self) -> None:
         url = './wiki/Special:EntityData/{}.json'.format(self.id)
         result = self.client.request(url)
+        if result is None:
+            self.data = None
```
Comment (dahlia) [inline wikidata/entity.py:241]: The existing code assumes if `self.data is None` it means entity data is not loaded *yet*, not non-existent.  So this code leads any `Entity` objects referring to non-existent entities to send a HTTP request again and again for every property/method call towards them.

Instead, we need three distinct types for three different cases:

1. `None`: Not-loaded yet
2. `Mapping[str, object]`: Existent and loaded
3. (to be chosen): Non-existent

**Comment by k----n** [comment]
@dahlia Implemented your suggested changes if non-existent entities, all checks pass

File: wikidata/client.py [hunk 7]
```diff
@@ -179,19 +180,28 @@ def decode_datavalue(self,
...
         if result is None:
             logger.debug('%r: no cache; make a request...', url)
-            response = self.opener.open(url)
-            buffer_ = io.TextIOWrapper(response, encoding='utf-8')
+            try:
+                response = self.opener.open(url)
+            except urllib.error.HTTPError as e:
+                logger.debug('HTTP error code: %s', e.code, exc_info=True)
+                response = e.fp.read().decode('utf-8')  # type: ignore
+                if 'Invalid ID' in response:
```
Comment (dahlia) [inline wikidata/client.py:198]: -  It would be better if it checks `e.code == 400`.

 -  Decoding `bytes` to `str` seems unnecessary.  `bytes` also supports `in` operator.

 -  [According to the docs][1]:

    > Though being an exception (a subclass of [`URLError`][2]), an [`HTTPError`][1] can also function as a non-exceptional file-like return value (the same thing that [`urlopen()`][3] returns).

    There you can write like `e.read()` instead of `e.fp.read()`.

Long story short:

```suggestion
                if e.code == 400 and b'Invalid ID' in e.read():
```

[1]: https://docs.python.org/3/library/urllib.error.html#urllib.error.HTTPError
[2]: https://docs.python.org/3/library/urllib.error.html#urllib.error.URLError
[3]: https://docs.python.org/3/library/urllib.request.html#urllib.request.urlopen

File: wikidata/client.py [hunk 8]
```diff
@@ -87,7 +88,7 @@ def __init__(self,
                  opener: Optional[urllib.request.OpenerDirector] = None,
                  datavalue_decoder: Union['Decoder',
                                           Callable[['Client', str,
-                                                   Mapping[str, object]],
+                                                    Mapping[str, object]],
```
Comment (dahlia) [inline wikidata/client.py:91]: Is this style change functionally necessary?  Or does lint require this change?

File: tests/client_test.py [hunk 9]
```diff
@@ -8,7 +8,7 @@
 from wikidata.entity import Entity, EntityId, EntityType
 
 if TYPE_CHECKING:
-    from typing import Dict, Union  # noqa: F401
+    from typing import Dict, Union   # noqa: F401
```
Comment (dahlia) [inline tests/client_test.py:11]: Is this style change functionally necessary?  Or does lint require this change?

File: wikidata/client.py [hunk 10]
```diff
@@ -179,19 +180,28 @@ def decode_datavalue(self,
...
     def request(self, path: str) -> Union[
         bool, int, float, str,
-        Mapping[
-            str,
-            Union[bool, int, float, str, Mapping[str, object], Sequence]
-        ],
-        Sequence[Union[bool, int, float, str, Mapping[str, object], Sequence]]
+        Mapping[str, Union[bool, int, float, str,
+                           Mapping[str, object], Sequence]],
+        Sequence[Union[bool, int, float, str, Mapping[str, object], Sequence]],
```
Comment (dahlia) [inline wikidata/client.py:185]: Is this style change functionally necessary?  Or does lint require this change?

File: wikidata/client.py [hunk 11]
```diff
@@ -179,19 +180,28 @@ def decode_datavalue(self,
...
             logger.debug('%r: no cache; make a request...', url)
-            response = self.opener.open(url)
-            buffer_ = io.TextIOWrapper(response, encoding='utf-8')
+            try:
+                response = self.opener.open(url)
+            except urllib.error.HTTPError as e:
+                logger.debug('HTTP error code: %s', e.code, exc_info=True)
+                response = e.fp.read().decode('utf-8')  # type: ignore
+                if 'Invalid ID' in response:
+                    return 'Non-Existent'
```
Comment (dahlia) [inline wikidata/client.py:199]: You added `None` to the arguments of `Union[...]` which is its return type.  So didn't you intend to return `None` for non-existent entities?

File: docs/changes.rst [hunk 12]
```diff
@@ -10,6 +10,7 @@ To be released.
 - Fixed :exc:`KeyError` from :meth:`Entity.getlist()
   <wikidata.entity.Entity.getlist>` if the property is explictly associated
   with "no value". [:issue:`18`]
+- Better handling of non-existent entities
```
Comment (dahlia) [inline docs/changes.rst:13]: Could you elaborate this?  Like, “what method became to return what when something does/is something,” blah blah.

File: wikidata/entity.py [hunk 13]
```diff
@@ -238,6 +238,13 @@ def attributes(self) -> Mapping[str, object]:
     def load(self) -> None:
         url = './wiki/Special:EntityData/{}.json'.format(self.id)
         result = self.client.request(url)
+        if result is None:
+            self.data = None  # Not loaded yet
+            return
+        elif result == 'Non-Existent':
+            self.data = {}
```
Comment (dahlia) [inline wikidata/entity.py:245]: I believe there would be a better way to represent non-existent entities than an empty dictionary.  How about `Entity` having an attribute/property like `.state`?  We can define an `enum` type for `Entity.state` like:

```python
class EntityState(enum.Enum):
    not_loaded = 'not_loaded'
    loaded = 'loaded'
    non_existent = 'non_existent'
```

File: wikidata/client.py (hunk 10 above)
Comment (k----n) [reply wikidata/client.py:185]: Linting otherwise line is too long

File: docs/changes.rst [hunk 14]
```diff
@@ -10,6 +10,7 @@ To be released.
 - Fixed :exc:`KeyError` from :meth:`Entity.getlist()
   <wikidata.entity.Entity.getlist>` if the property is explictly associated
   with "no value". [:issue:`18`]
+- Added Enum to handle non-existent entities
```
Comment (dahlia) [inline docs/changes.rst:13]:
```suggestion
- Added :class:`~wikidata.entity.EntityState` enum class.  [:pr:`11`]
- Added :attr:`Entity.state <wikidata.entity.Entity.state>` attribute.
  [:pr:`11`]
- Fixed a bug that raised :exc:`~urllib.error.HTTPError` when
  non-existent `Entity` was requested.  [:pr:`11`]
```

File: wikidata/entity.py [hunk 15]
```diff
@@ -238,6 +247,14 @@ def attributes(self) -> Mapping[str, object]:
     def load(self) -> None:
         url = './wiki/Special:EntityData/{}.json'.format(self.id)
         result = self.client.request(url)
+        if result is None:
+            self.data = None  # Not loaded yet
+            self.state = EntityState.not_loaded
+            return
+        elif result == 'Non-Existent':
+            self.state = EntityState.non_existent
+            return
```
Comment (dahlia) [inline wikidata/entity.py:256]: `Client.request()` returning `None` does not mean “not loaded yet” but “not able to load (non-existent).”

File: wikidata/entity.py [hunk 16]
```diff
@@ -238,6 +247,14 @@ def attributes(self) -> Mapping[str, object]:
     def load(self) -> None:
```
Comment (dahlia) [inline wikidata/entity.py:247]: It should return early if `self.state is Entity.non_existent`, because `Entity.non_existent` means the server will respond with nothing anyway.

File: wikidata/entity.py [hunk 17]
```diff
@@ -253,6 +270,7 @@ def load(self) -> None:
         assert isinstance(data, collections.abc.Mapping)
         self.data = data
         self.id = entity_id
+        self.state = EntityState.loaded
```
Comment (dahlia) [inline wikidata/entity.py:273]: What does this line purpose to?

Comment (k----n) [reply wikidata/entity.py:273]: Indicates entity data has been loaded

File: tests/mock.py [hunk 18]
```diff
@@ -12,8 +12,8 @@
 import urllib.request
 import urllib.response
 
-__all__ = ('ENTITY_FIXTURES_PATH', 'FIXTURES_PATH', 'MEDIA_FIXTURES_PATH',
-           'FixtureOpener')
+__all__ = ['ENTITY_FIXTURES_PATH', 'FIXTURES_PATH', 'MEDIA_FIXTURES_PATH',
+           'FixtureOpener']
```
Comment (dahlia) [inline tests/mock.py:16]: Please do not change the existing style.

File: wikidata/cache.py [hunk 19]
```diff
@@ -11,8 +11,8 @@
 import re
 from typing import NewType, Optional
 
-__all__ = ('CacheKey', 'CachePolicy', 'CacheValue',
-           'MemoryCachePolicy', 'NullCachePolicy', 'ProxyCachePolicy')
+__all__ = ['CacheKey', 'CachePolicy', 'CacheValue',
+           'MemoryCachePolicy', 'NullCachePolicy', 'ProxyCachePolicy']
```
Comment (dahlia) [inline wikidata/cache.py:15]: Please do not change the existing style.

File: wikidata/client.py [hunk 20]
```diff
@@ -18,7 +19,7 @@
 if TYPE_CHECKING:
     from .datavalue import Decoder  # noqa: F401
 
-__all__ = 'WIKIDATA_BASE_URL', 'Client'
+__all__ = ['WIKIDATA_BASE_URL', 'Client']
```
Comment (dahlia) [inline wikidata/client.py:22]: Please do not change the existing style.

File: wikidata/commonsmedia.py [hunk 21]
```diff
@@ -12,7 +12,7 @@
 
 from .client import Client
 
-__all__ = 'File', 'FileError'
+__all__ = ['File', 'FileError']
```
Comment (dahlia) [inline wikidata/commonsmedia.py:15]: Please do not change the existing style.

File: wikidata/datavalue.py [hunk 22]
```diff
@@ -24,7 +24,7 @@
 if TYPE_CHECKING:
     from .entity import Entity  # noqa: F401
 
-__all__ = 'DatavalueError', 'Decoder'
+__all__ = ['DatavalueError', 'Decoder']
```
Comment (dahlia) [inline wikidata/datavalue.py:27]: Please do not change the existing style.

File: wikidata/entity.py [hunk 23]
```diff
@@ -17,7 +17,7 @@
 if TYPE_CHECKING:
     from .client import Client  # noqa: F401
 
-__all__ = 'Entity', 'EntityId', 'EntityType'
+__all__ = ['Entity', 'EntityId', 'EntityState', 'EntityType']
```
Comment (dahlia) [inline wikidata/entity.py:20]: Please leave it as a `tuple`.

```suggestion
__all__ = 'Entity', 'EntityId', 'EntityState', 'EntityType'
```

File: wikidata/multilingual.py [hunk 24]
```diff
@@ -7,7 +7,7 @@
 
 from babel.core import Locale
 
-__all__ = 'MonolingualText', 'MultilingualText', 'normalize_locale_code'
+__all__ = ['MonolingualText', 'MultilingualText', 'normalize_locale_code']
```
Comment (dahlia) [inline wikidata/multilingual.py:10]: Please do not change the existing style.

File: wikidata/client.py (hunk 10 above)
Comment (dahlia) [reply wikidata/client.py:185]: Fair enough.

**Comment by k----n** [comment]
> Thanks for your contribution!

Thanks for the guidance on better code! Much appreciated.
